### PR TITLE
feat: more Gateway API kind wrappers

### PR DESCRIPTION
**Pull request by KevFan**
# Description
Closes: https://github.com/Kuadrant/policy-machinery/issues/11

- Adds wrappers the the following so they can be used in the topology as targetable / object / policy:
  - **gateway.networking.k8s.io/v1**
    - `GRPCRoute` (targetable)
  
  - **gateway.networking.k8s.io/v1beta1**
    - `ReferenceGrant` (object)
  
  - **gateway.networking.k8s.io/v1alpha3** 
    - `BackendTLSPolicy` (policy)
  
  - **gateway.networking.k8s.io/v1alpha2** 
    - `TCPRoute` (targetable)
    - `TLSRoute` (targetable)
    - `UDPRoute` (targetable)
    - `BackendLBPolicy` (policy)
  
  - **v1 (core)**
    - `Namespace` (targetable)

- Additionally adds the `LinkFunc`'s for the new `xRoute` kinds from above as part of the Gateway API Topology
  - These were **not** added to `controller/topology_builder` for now, unless we want to enable these already as part of the controller package?
  -  The link function for `Namespace` as a targetable in the topolgy was **not**  added as part of this PR as this seems not as straightforward which might require some deeper considerations such as:
     - Will `Namespace` be also a parent to `Gateway` since `Gateway` is namespace scoped? (i.e. `Gateway` will have both `GatewayClass` and `Namespace` as parents ). Who will take precedence in this case?
     - `Namespace` can be a parent of whatever the resource in its namespace which are also targetable, does this matter? i.e. a Route in a namespace may not have a Gateway since a Route can use a Gateway from a different namespace as a parent reference

one_of_each_kind

complex_not_expanded

with_section_name

complex

### CODE REVIEW

File: machinery/gateway_api_topology_test.go [hunk 1]
```diff
@@ -20,6 +20,10 @@ import (
 // This results in a topology with the following scheme:
 //
 //	GatewayClass -> Gateway -> HTTPRoute -> Service
+//							∟> GRPCRoute ⤴
+//							∟> TCPRoute  ⤴
+//							∟> TLSRoute  ⤴
+//							∟> UDPRoute  ⤴
```
Comment (guicassolato) [inline machinery/gateway_api_topology_test.go:26]: I guess some white spaces have been converted to tabs here.

```suggestion
//	                        ∟> GRPCRoute ⤴
//	                        ∟> TCPRoute  ⤴
//	                        ∟> TLSRoute  ⤴
//	                        ∟> UDPRoute  ⤴
```

File: machinery/core_types.go [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+	return UrlFromObject(n)
+}
+
+func (n *Namespace) SetPolicies(policies []Policy) {
+	n.attachedPolicies = policies
+}
+
+func (n *Namespace) Policies() []Policy {
+	return n.attachedPolicies
+}
```
Comment (guicassolato) [inline machinery/core_types.go:29]: I guess `Namespace` is a special case, which we need to think in the context of the hierarchy of network resources.

This is a targetable that every other targetable the lib provides a wrapper for (with the exception of GatewayClass) can be linked to. The link itself is pretty straightforward and probably we could provide the helpers. But then we need to think what a link means in terms of inheritance.

E.g. a HTTPRoute is linked to a Gateway (via `spec.parentRef`) and to a Namespace (via `metadata.namespace`). In a scenario where we have policies attached to each of these resources, the final effective policy at the level of that one HTTPRoute will be composed out of `Gateway policies → HTTPRoute Namespace policies → HTTPRoute policies`. However, for an occasional second HTTPRoute in the same namespace, it doesn't mean necessarily the Gateway policies will have any effect, because it also depends if the second HTTPRoute is parented by that Gateway or not. Furthermore, the Gateway itself could be in the same namespace of the routes. I wonder what happens to the hierarchy of policies in this case.

One idea I had was of defining wrappers for namespace for each kind of namespaced targetable, i.e. `GatewayNamespace`, `HTTPRouteNamespace`, etc. But we would not be able to resolve the Policy → Namespace links from the targetRef only. So we need to think this through again.

File: machinery/gateway_api_topology.go [hunk 3]
```diff
@@ -8,19 +8,46 @@ import (
...
+	GRPCRouteGroupKind     = schema.GroupKind{Group: gwapiv1.GroupVersion.Group, Kind: "GRPCRoute"}
+	GRPCRouteRuleGroupKind = schema.GroupKind{Group: gwapiv1.GroupVersion.Group, Kind: "GRPCRouteRule"}
+	TCPRouteGroupKind      = schema.GroupKind{Group: gwapiv1alpha2.GroupVersion.Group, Kind: "TCPRoute"}
+	TCPRouteRuleGroupKind  = schema.GroupKind{Group: gwapiv1alpha2.GroupVersion.Group, Kind: "TCPRouteRule"}
+	TLSRouteGroupKind      = schema.GroupKind{Group: gwapiv1alpha2.GroupVersion.Group, Kind: "TLSRoute"}
+	TLSRouteRuleGroupKind  = schema.GroupKind{Group: gwapiv1alpha2.GroupVersion.Group, Kind: "TLSRouteRule"}
+	UDPRouteGroupKind      = schema.GroupKind{Group: gwapiv1alpha2.GroupVersion.Group, Kind: "UDPRoute"}
+	UDPRouteRuleGroupKind  = schema.GroupKind{Group: gwapiv1alpha2.GroupVersion.Group, Kind: "UDPRouteRule"}
+	ServiceGroupKind       = schema.GroupKind{Kind: "Service"}
+	ServicePortGroupKind   = schema.GroupKind{Kind: "ServicePort"}
```
Comment (guicassolato) [inline machinery/gateway_api_topology.go:29]: We've removed the wrapper from "Gateway API types" now into "Core types". Should we define these there instead?

I also wonder if this should in the `*types.go` file or in `*topology.go`.

File: machinery/gateway_api_types.go [hunk 4]
```diff
@@ -315,6 +499,82 @@ func (t LocalPolicyTargetReferenceWithSectionName) GetName() string {
 	return namespacedSectionName(string(t.LocalPolicyTargetReference.Name), *t.SectionName)
 }
 
+// These are wrappers for Gateway API types so instances can be used as objects in the topology.
+
+type ReferenceGrant struct {
```
Comment (guicassolato) [inline machinery/gateway_api_types.go:504]: Could the lib provide link func helpers?

File: controller/topology_builder.go [hunk 5]
```diff
@@ -24,31 +24,31 @@ type gatewayAPITopologyBuilder struct {
 }
 
 func (t *gatewayAPITopologyBuilder) Build(objs Store) *machinery.Topology {
-	gatewayClasses := lo.FilterMap(lo.Values(objs[schema.GroupKind{Group: gwapiv1.GroupVersion.Group, Kind: "GatewayClass"}]), func(obj RuntimeObject, _ int) (*gwapiv1.GatewayClass, bool) {
+	gatewayClasses := lo.FilterMap(lo.Values(objs[machinery.GatewayClassGroupKind]), func(obj RuntimeObject, _ int) (*gwapiv1.GatewayClass, bool) {
 		gc, ok := obj.(*gwapiv1.GatewayClass)
 		if !ok {
 			return nil, false
 		}
 		return gc, true
```
Comment (guicassolato) [inline controller/topology_builder.go:32]: It doesn't belong to this PR exactly, but I guess these filter maps could be simplified.

```suggestion
	gatewayClasses := lo.FilterMap(lo.Values(objs[machinery.GatewayClassGroupKind]), func(obj RuntimeObject, _ int) (*gwapiv1.GatewayClass, bool) {
		return obj.(*gwapiv1.GatewayClass)
```

File: controller/topology_builder.go [hunk 6]
```diff
@@ -24,31 +24,31 @@ type gatewayAPITopologyBuilder struct {
...
+	httpRoutes := lo.FilterMap(lo.Values(objs[machinery.HTTPRouteGroupKind]), func(obj RuntimeObject, _ int) (*gwapiv1.HTTPRoute, bool) {
 		httpRoute, ok := obj.(*gwapiv1.HTTPRoute)
 		if !ok {
 			return nil, false
 		}
 		return httpRoute, true
 	})
 
-	services := lo.FilterMap(lo.Values(objs[schema.GroupKind{Group: core.GroupName, Kind: "Service"}]), func(obj RuntimeObject, _ int) (*core.Service, bool) {
+	services := lo.FilterMap(lo.Values(objs[machinery.ServiceGroupKind]), func(obj RuntimeObject, _ int) (*core.Service, bool) {
```
Comment (guicassolato) [inline controller/topology_builder.go:51]: Let's also add the new types of resources here?

File: machinery/core_types.go (hunk 2 above)
Comment (KevFan) [reply machinery/core_types.go:29]: Yes, exactly, `Namespace` looks to be a special case as a `Targetable` and what it means in terms of inheritance for the effective policy. For now, I was thinking of just adding a wrapper for this as part of this PR. The link functions can be provided as a seperate PR after this is thought through

File: machinery/gateway_api_topology.go (hunk 3 above)
Comment (KevFan) [reply machinery/gateway_api_topology.go:29]: Moving to `*types.go` makes sense to me, although I've no strong preferance

File: controller/topology_builder.go (hunk 5 above)
Comment (guicassolato) [reply controller/topology_builder.go:32]: Done for the existing wrappers in https://github.com/Kuadrant/policy-machinery/pull/17/commits/112499121c28f3700e4234e1f0413bd76473043d#diff-cf66b18260b0367d1ba0a51dd11004e8d9a5ec5c37c31c92be70630305bc7737

Comment (KevFan) [reply controller/topology_builder.go:32]: The above gives an error `not enough return values` on the IDE, so the above won't directly unless I'm missing something 🤔 

The below would work I think without hardcoding the boolean returns like `return obj.(*gwapiv1.GatewayClass) false|true`.  
```suggestion
	gatewayClasses := lo.FilterMap(lo.Values(objs[machinery.GatewayClassGroupKind]), func(obj RuntimeObject, _ int) (*gwapiv1.GatewayClass, bool) {
		gc, ok := obj.(*gwapiv1.GatewayClass)
		return gc, ok
```
What do you think?

Comment (guicassolato) [reply controller/topology_builder.go:32]: You're right. We need to lines at least.

File: machinery/core_types.go (hunk 2 above)
Comment (KevFan) [reply machinery/core_types.go:35]: I've added a linking functions for `Namespace -> Gateway ` and  `Namespace -> xRoute` that looks like this at the moment:

image

If a namespace contains a route but parent refs a gateway in a different namespace, there is a direct link the the route to the namespace. Otherwise if the route parent ref gateway is in the same namespace as the route, this link from `Namespace -> Route` is skipped as it is already linked by `Namespace -> Gateway -> Route`.

Let me know your thought on this

File: machinery/gateway_api_types.go (hunk 4 above)
Comment (KevFan) [reply machinery/gateway_api_types.go:520]: I've been thinking on this but I'm a bit confused what kind of links funcs for `ReferenceGrants` to do 😅 

Are we linking everything under `referencegrant.spec.from` and `referencegrant.spec.to` to the `RefernceGrant` or are we just linking `referencegrant.spec.from` to `referencegrant.spec.to` or are you thinking something else? 🤔

File: machinery/core_types.go (hunk 2 above)
Comment (guicassolato) [reply machinery/core_types.go:35]: > if the route parent ref gateway is in the same namespace as the route, this link from `Namespace -> Route` is skipped as it is already linked by `Namespace -> Gateway -> Route`.

I understand this would make for a clean graph, but it seems to add complexity IMO. I guess I would keep it simple: if resource _R_ is in namespace _N_, then `N → R`.

Moreover, conceptually the relationship between namespace and route is not transitive via the gateway, but a direct one. It exists regardless of the existence of the gateway. Users should get the route _and_ gateway when asking for the child targetables of a namespace if these resources are both in the namespace.

IDK... to me it looks like we're better off without the special case. But if you've put some thought on this. There may be something I'm missing perhaps.

Comment (KevFan) [reply machinery/core_types.go:34]: We've decided to move Namespace linking to a separate PR

File: machinery/gateway_api_types.go (hunk 4 above)
Comment (KevFan) [reply machinery/gateway_api_types.go:520]: We've decided ReferanceGrant linking to a seperate PR

**Comment by KevFan** [comment]
Still looking into the unit test failure, it's failing locally for me now. Seems to an issue that surfaced after the new dot engine PR was merged. `Service -> ServicePort` is linked multiple times now on the graph for some reason 🤔 

image

File: machinery/gateway_api_topology.go [hunk 7]
```diff
@@ -170,6 +256,110 @@ func NewGatewayAPITopology(options ...GatewayAPITopologyOptionsFunc) *Topology {
...
 	}
 
+	if o.ExpandGRPCRouteRules {
+		grpcRouteRules := lo.FlatMap(o.GRPCRoutes, GRPCRouteRulesFromGRPCRouteRule)
+		opts = append(opts, WithTargetables(grpcRouteRules...))
+		opts = append(opts, WithLinks(LinkGRPCRouteToGRPCRouteRuleFunc())) // GRPCRoute -> GRPCRouteRule
+
+		if o.ExpandServicePorts {
+			servicePorts := lo.FlatMap(o.Services, ServicePortsFromBackendFunc)
+			opts = append(opts, WithTargetables(servicePorts...))
```
Comment (guicassolato) [inline machinery/gateway_api_topology.go:266]: This creates the ServicePorts again, after having created within the expansion of HTTPRouteRules. We don't realise that because the URL of the object is the same.

We probably need to rewrite this logic to expand ServicePorts independently from the expansion of route rules, and just link accordingly. IOW, move these two lines out of the expansion of the route rules.

File: machinery/gateway_api_topology.go [hunk 8]
```diff
@@ -170,6 +256,110 @@ func NewGatewayAPITopology(options ...GatewayAPITopologyOptionsFunc) *Topology {
...
+		}
+	}
+
+	if o.ExpandTCPRouteRules {
+		tcpRouteRules := lo.FlatMap(o.TCPRoutes, TCPRouteRulesFromTCPRouteFunc)
+		opts = append(opts, WithTargetables(tcpRouteRules...))
+		opts = append(opts, WithLinks(LinkTCPRouteToTCPRouteRuleFunc())) // TCPRoute - TCPRouteRules
+
+		if o.ExpandServicePorts {
+			servicePorts := lo.FlatMap(o.Services, ServicePortsFromBackendFunc)
```
Comment (guicassolato) [inline machinery/gateway_api_topology.go:291]: Do you mind please renaming this mapping func? In a previous PR I renamed "Backend" to "Service", but missed this one.

```suggestion
			servicePorts := lo.FlatMap(o.Services, ServicePortsFromServiceFunc)
```

File: machinery/gateway_api_topology.go [hunk 9]
```diff
@@ -170,6 +256,110 @@ func NewGatewayAPITopology(options ...GatewayAPITopologyOptionsFunc) *Topology {
...
+				LinkUDPRouteToServicePortFunc(o.UDPRoutes),   // UDPRoute -> ServicePort
+				LinkUDPRouteToServiceFunc(o.UDPRoutes, true), // UDPRoute -> Service
+			))
+		} else {
+			opts = append(opts, WithLinks(LinkUDPRouteToServiceFunc(o.UDPRoutes, false))) // UDPRoute -> Service
+		}
+	}
+
 	if o.ExpandServicePorts {
 		opts = append(opts, WithLinks(LinkServiceToServicePortFunc())) // Service -> ServicePort
```
Comment (guicassolato) [inline machinery/gateway_api_topology.go:364]: Related to https://github.com/Kuadrant/policy-machinery/pull/16/files#r1719696707

```suggestion
		servicePorts := lo.FlatMap(o.Services, ServicePortsFromBackendFunc)
		opts = append(opts, WithTargetables(servicePorts...))
		opts = append(opts, WithLinks(LinkServiceToServicePortFunc())) // Service -> ServicePort
```

File: machinery/gateway_api_topology.go (hunk 7 above)
Comment (KevFan) [reply machinery/gateway_api_topology.go:266]: Oh, yes, that's it, thanks 👍

File: machinery/gateway_api_topology.go (hunk 8 above)
Comment (KevFan) [reply machinery/gateway_api_topology.go:291]: Sure, I'll rename this 👍

File: machinery/core_types.go (hunk 2 above)
Comment (Boomatang) [reply machinery/core_types.go:34]: If the Namespace linking is in a a separate PR, should the adding of the wrapper also be in that PR?

Comment (KevFan) [reply machinery/core_types.go:34]: I think it's fine / safe to keep the wrapper in this PR as it was an original ask of the issue

Comment (Boomatang) [reply machinery/core_types.go:34]: I am wondering is this not effectively dead code currently. I know it is safe to have it there but what use does it give a user of the library. I am looking at this from the point it is harder to remove code than add it in later. If it is to be left there, sure I am okay with that but if there is no use case of it today I think we should remove it.

Comment (KevFan) [reply machinery/core_types.go:34]: I guess the only use now by having the wrapper is that users of this library can just use this struct while providing their own link functions if they so choose (until we provide the link functions, but even then, this is all opt-in depending on how they want to create their topology). Without this, a user would also need to wrap the `Namespace` struct themselves with pretty much the same implementation (unless they want to implement it differently) 🤔 

I've no strong preference of this either, but if we do decide to remove this until later, I'll also remove the `ReferenceGrant` wrapper, since the same will apply to it

Comment (guicassolato) [reply machinery/core_types.go:34]: > users of this library can just use this struct while providing their own link functions

Agreed. This may also be a good way to learn the best way to provide the built-in link funcs later on.

File: machinery/core_types.go [hunk 10]
```diff
@@ -0,0 +1,91 @@
+package machinery
+
+import (
+	core "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	gwapiv1 "sigs.k8s.io/gateway-api/apis/v1"
+)
+
+var (
+	ServiceGroupKind     = core.SchemeGroupVersion.WithKind("Service").GroupKind()
```
Comment (guicassolato) [inline machinery/core_types.go:10]: Maybe add it for Namespace too?

```suggestion
	NamespaceGroupKind   = core.SchemeGroupVersion.WithKind("Namespace").GroupKind()
	ServiceGroupKind     = core.SchemeGroupVersion.WithKind("Service").GroupKind()
```

File: machinery/gateway_api_types.go [hunk 11]
```diff
@@ -3,11 +3,29 @@ package machinery
...
+	HTTPRouteGroupKind     = gwapiv1.SchemeGroupVersion.WithKind("HTTPRoute").GroupKind()
+	HTTPRouteRuleGroupKind = gwapiv1.SchemeGroupVersion.WithKind("HTTPRouteRule").GroupKind()
+	GRPCRouteGroupKind     = gwapiv1.SchemeGroupVersion.WithKind("GRPCRoute").GroupKind()
+	GRPCRouteRuleGroupKind = gwapiv1.SchemeGroupVersion.WithKind("GRPCRouteRule").GroupKind()
+	TCPRouteGroupKind      = gwapiv1alpha2.SchemeGroupVersion.WithKind("TCPRoute").GroupKind()
+	TCPRouteRuleGroupKind  = gwapiv1alpha2.SchemeGroupVersion.WithKind("TCPRouteRule").GroupKind()
+	TLSRouteGroupKind      = gwapiv1alpha2.SchemeGroupVersion.WithKind("TLSRoute").GroupKind()
+	TLSRouteRuleGroupKind  = gwapiv1alpha2.SchemeGroupVersion.WithKind("TLSRouteRule").GroupKind()
+	UDPRouteGroupKind      = gwapiv1alpha2.SchemeGroupVersion.WithKind("UDPRoute").GroupKind()
+	UDPRouteRuleGroupKind  = gwapiv1alpha2.SchemeGroupVersion.WithKind("UDPRouteRule").GroupKind()
```
Comment (guicassolato) [inline machinery/gateway_api_types.go:28]: ReferenceGrant, BackendTLSPolicy and BackendLBPolicy.

```suggestion
	GatewayClassGroupKind      = gwapiv1.SchemeGroupVersion.WithKind("GatewayClass").GroupKind()
	GatewayGroupKind           = gwapiv1.SchemeGroupVersion.WithKind("Gateway").GroupKind()
	ListenerGroupKind          = gwapiv1.SchemeGroupVersion.WithKind("Listener").GroupKind()
	HTTPRouteGroupKind         = gwapiv1.SchemeGroupVersion.WithKind("HTTPRoute").GroupKind()
	HTTPRouteRuleGroupKind     = gwapiv1.SchemeGroupVersion.WithKind("HTTPRouteRule").GroupKind()
	GRPCRouteGroupKind         = gwapiv1.SchemeGroupVersion.WithKind("GRPCRoute").GroupKind()
	GRPCRouteRuleGroupKind     = gwapiv1.SchemeGroupVersion.WithKind("GRPCRouteRule").GroupKind()
	ReferenceGrantGroupKind    = gwapiv1beta1.SchemeGroupVersion.WithKind("ReferenceGrant").GroupKind()
	BackendTLSPolicyGroupKind  = gwapiv1alpha3.SchemeGroupVersion.WithKind("BackendTLSPolicy").GroupKind()
	BackendLBPolicyGroupKind   = gwapiv1alpha2.SchemeGroupVersion.WithKind("BackendLBPolicy").GroupKind()
	TCPRouteGroupKind          = gwapiv1alpha2.SchemeGroupVersion.WithKind("TCPRoute").GroupKind()
	TCPRouteRuleGroupKind      = gwapiv1alpha2.SchemeGroupVersion.WithKind("TCPRouteRule").GroupKind()
	TLSRouteGroupKind          = gwapiv1alpha2.SchemeGroupVersion.WithKind("TLSRoute").GroupKind()
	TLSRouteRuleGroupKind      = gwapiv1alpha2.SchemeGroupVersion.WithKind("TLSRouteRule").GroupKind()
	UDPRouteGroupKind          = gwapiv1alpha2.SchemeGroupVersion.WithKind("UDPRoute").GroupKind()
	UDPRouteRuleGroupKind      = gwapiv1alpha2.SchemeGroupVersion.WithKind("UDPRouteRule").GroupKind()
```

**Review by guicassolato — APPROVED** [review]
Left a minor suggestion only for completeness, but excellent work overall. Thanks!
